box_error.go

```diff
@@ -0,0 +1,289 @@
+							return nil, err
+						}
+						for ; l2 > 0; l2-- {
+							if mapk, err = d.DecodeString(); err != nil {
+								return nil, err
+							}
+							if mapv, err = d.DecodeInterface(); err != nil {
+								return nil, err
+							}
+							errorStack[i].Fields[mapk] = mapv
+						}
+					default:
+						if err = d.Skip(); err != nil {
+							return nil, err
+						}
+					}
+				}
+
+				if i > 0 {
+					errorStack[i-1].Prev = &errorStack[i]
+				}
+			}
+		default:
+			if err = d.Skip(); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if len(errorStack) > 0 {
+		return &errorStack[0], nil
+	}
+
+	return nil, nil
+}
+
+func encodeBoxError(enc *encoder, boxError *BoxError) error {
+	if err := enc.EncodeMapLen(1); err != nil {
+		return err
+	}
+	if err := encodeUint(enc, keyErrorStack); err != nil {
+		return err
+	}
+
+	var stackDepth = boxError.Depth()
+	if err := enc.EncodeArrayLen(stackDepth); err != nil {
+		return err
+	}
+
+	for ; stackDepth > 0; stackDepth-- {
+		fieldsLen := len(boxError.Fields)
+
+		if fieldsLen > 0 {
+			if err := enc.EncodeMapLen(7); err != nil {
+				return err
+			}
+		} else {
+			if err := enc.EncodeMapLen(6); err != nil {
+				return err
+			}
+		}
+
+		if err := encodeUint(enc, keyErrorType); err != nil {
+			return err
+		}
+		if err := enc.EncodeString(boxError.Type); err != nil {
+			return err
+		}
+
+		if err := encodeUint(enc, keyErrorFile); err != nil {
+			return err
+		}
+		if err := enc.EncodeString(boxError.File); err != nil {
+			return err
+		}
+
+		if err := encodeUint(enc, keyErrorLine); err != nil {
+			return err
+		}
+		if err := enc.EncodeUint64(boxError.Line); err != nil {
+			return err
+		}
+
+		if err := encodeUint(enc, keyErrorMessage); err != nil {
+			return err
+		}
+		if err := enc.EncodeString(boxError.Msg); err != nil {
+			return err
+		}
+
+		if err := encodeUint(enc, keyErrorErrno); err != nil {
+			return err
+		}
+		if err := enc.EncodeUint64(boxError.Errno); err != nil {
+			return err
+		}
+
+		if err := encodeUint(enc, keyErrorErrcode); err != nil {
+			return err
+		}
+		if err := enc.EncodeUint64(boxError.Code); err != nil {
+			return err
+		}
+
+		if fieldsLen > 0 {
+			if err := encodeUint(enc, keyErrorFields); err != nil {
+				return err
+			}
+
+			if err := enc.EncodeMapLen(fieldsLen); err != nil {
+				return err
+			}
+
+			for k, v := range boxError.Fields {
+				if err := enc.EncodeString(k); err != nil {
+					return err
+				}
+				if err := enc.Encode(v); err != nil {
+					return err
+				}
+			}
+		}
+
+		if stackDepth > 1 {
+			boxError = boxError.Prev
+		}
+	}
+
+	return nil
+}
+
+// UnmarshalMsgpack deserializes a BoxError value from a MessagePack
+// representation.
+func (e *BoxError) UnmarshalMsgpack(b []byte) error {
+	var val *BoxError
+	var err error
+
+	buf := bytes.NewBuffer(b)
+	dec := newDecoder(buf)
+
+	if val, err = decodeBoxError(dec); err != nil {
+		return err
+	}
+
+	if val != nil {
+		if e == nil {
+			e = &BoxError{}
+		}
```

Comment:
It doesn't seem to make sense. Will we set a value to a stack-local varibale at the next line?

Reply:
Yeah, it seems I need to replace this one with `make`.

Reply:
Wait a second, `make` is a tool to create slices.

Based on [this](https://stackoverflow.com/questions/31786937/are-all-the-variables-in-go-allocated-on-heap) stackoverflow answer, this approach seems ok:

> [There's a very clear answer to that question in the FAQ](https://golang.org/doc/faq#stack_or_heap):
> 
>     How do I know whether a variable is allocated on the heap or the stack?
> 
>     From a correctness standpoint, you don't need to know. Each variable in Go exists as long as there are references to it. The storage location chosen by the implementation is irrelevant to the semantics of the language.
> 
>     The storage location does have an effect on writing efficient programs. When possible, the Go compilers will allocate variables that are local to a function in that function's stack frame. However, if the compiler cannot prove that the variable is not referenced after the function returns, then the compiler must allocate the variable on the garbage-collected heap to avoid dangling pointer errors. Also, if a local variable is very large, it might make more sense to store it on the heap rather than the stack.
> 
>     In the current compilers, if a variable has its address taken, that variable is a candidate for allocation on the heap. However, a basic escape analysis recognizes some cases when such variables will not live past the return from the function and can reside on the stack.
> 
> TLDR: You shouldn't care. Go takes care of allocation for you.

Reply:
https://go.dev/play/p/rrQdeadMzPJ :

```Go
import "fmt"

type foo struct {
	val int
}

func (a *foo) foo() {
	if a == nil {
		fmt.Println("a == nil")
		a = &foo{0}
		*a = foo{1}
	}
}

func main() {
	var bar *foo
	bar.foo()
	fmt.Println(bar)
}
// Output:
// a == nil
// <nil>
```

Reply:
I'm confused. So how does one fills the pointer value?

Reply:
Or is it simply impossible with nil pointer? The question is "What should I do if one had called UnmarshalMsgpack for a nil pointer?" and before I had added this line of code the answer was "crash with segmentation fault".

Reply:
Reworked to panic for `nil` pointer.

Reply:
I think explicit panic is better, segfault is always dissatisfying.